File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        s1, s2 = text1, text2
        l1, l2 = len(text1), len(text2)
        row, col = l1+1, l2+1
        
        if not l1 or not l2:
            return 0
        
        memo = [0 for _ in range(col)]
        
        for r in range(1, row):
            # use prev to keep memo[r-1][c-1]
            prev = memo[0]
            for c in range(1, col):
                tmp = memo[c] 
                if s1[r-1] == s2[c-1]:
                    memo[c] = prev + 1
                else:
                    memo[c] = max(memo[c-1], memo[c])
                prev = tmp
            
        return memo[-1]

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
    string s1, s2;
    vector<vector<int>> dp;
public:
    int solve(int i, int j) {
        if(i >= s1.size() || j >= s2.size()) return 0;
        int ans = 0;
        if(dp[i][j] != -1) return dp[i][j];
        if(s1[i] == s2[j]) {
            ans += solve(i + 1, j + 1) + 1;
        } else
            ans += max(solve(i + 1, j), solve(i, j + 1));
        return dp[i][j] = ans;
    }
    int longestCommonSubsequence(string text1, string text2) {
        s1 = text1;
        s2 = text2;
        dp.resize(s1.size(), vector<int>(s2.size(), -1));
        return solve(0,0);
    }
};